### ISSUE TITLE
MS SQL Server Express Error on Load

### ISSUE THREAD

**Issue by joelschuster** [moqui/moqui-framework#206]
I am trying to use Moqui with MS SQL Server Express. I installed a docker container for SQL Server, the official docker image from Microsoft for linux.

**MS SQL Server Express Install**

_Docker Image:_
https://hub.docker.com/r/microsoft/mssql-server-windows-express/

_MS Instructions:_
https://docs.microsoft.com/en-us/sql/linux/sql-server-linux-setup-docker

**Configuration**
I then updated my MoquiDevConf.xml by adding the mssql datasource:

```
<entity-facade query-stats="true">
        <!-- add datasource elements here to configure databases -->
		<datasource group-name="transactional" database-conf-name="mssql" schema-name="">
            <inline-jdbc jdbc-uri="jdbc:sqlserver://192.168.175.129:1433;databaseName=moqui" jdbc-username="moqui" jdbc-password="moqui"/>
        </datasource>
    </entity-facade>
```

... where I was able to confirm the IP and port using both MS SQL Server Management Studio from my host machine and from Intellij via the JDBC drivers.

**JDBC Driver**
I also added the MS JDBC drivers 6.0 for MS SQL Server to the runtime/lib directory of moqui from here: 

https://www.microsoft.com/en-us/download/details.aspx?id=11774

**Moqui Load Attempt**

I get the following error when trying

`gradlew load`

(Log File is also attached)

```
09:13:36.259  WARN         main                o.m.i.e.EntityValueImpl Error creating [moqui.server.ArtifactHitBin: {hitBinId=100000, lastUpdatedStamp=2017-04-14 09:13:36.238, artifactType=AT_ENTITY, hitCount=1, artifactSubType=refresh, binStartDateTime=2017-04-14 09:11:55.188, minTimeMillis=12.7292450000000005871925168321467936038970947265625, serverHostName=joels-win, binEndDateTime=2017-04-14 09:13:36.22, totalSquaredTime=162.0336782700250068955938331782817840576171875, slowHitCount=0, artifactName=moqui.basic.EnumerationType, totalTimeMillis=12.7292450000000005871925168321467936038970947265625, serverIpAddress=192.168.1.102, maxTimeMillis=12.7292450000000005871925168321467936038970947265625}] tx [could not get] con a ConnectionJavaProxy of a JdbcPooledConnection from datasource transactional_DS in state ACCESSIBLE with usage count 1 wrapping a JDBC LrcXAConnection on a JDBC LrcConnectionJavaProxy on ConnectionID:1 ClientConnectionId: 92c6dacc-45ff-4976-b5d8-1e34726be4e1 on a JDBC LrcConnectionJavaProxy on ConnectionID:1 ClientConnectionId: 92c6dacc-45ff-4976-b5d8-1e34726be4e1: com.microsoft.sqlserver.jdbc.SQLServerException: Error converting data type nvarchar to decimal.
Error loading or running Moqui.loadData with args [{load=}]: java.lang.reflect.InvocationTargetException
java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at MoquiStart.main(MoquiStart.java:141)
Caused by: Error creating ArtifactHitBin {hitBinId=100000} [S0005]
        at org.moqui.impl.entity.EntityValueBase.create(EntityValueBase.java:1221)
        at org.moqui.impl.context.ExecutionContextFactoryImpl.destroy(ExecutionContextFactoryImpl.groovy:623)
        at org.moqui.Moqui.loadData(Moqui.java:154)
        ... 5 more
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: Error converting data type nvarchar to decimal.
        at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:217)
        at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1655)
        at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatement(SQLServerPreparedStatement.java:440)
        at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtExecCmd.doExecute(SQLServerPreparedStatement.java:385)
        at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7505)
        at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:2445)
        at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:191)
        at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:166)
        at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.executeUpdate(SQLServerPreparedStatement.java:328)
        at bitronix.tm.resource.jdbc.proxy.PreparedStatementJavassistProxy.executeUpdate(PreparedStatementJavassistProxy.java)
        at org.moqui.impl.entity.EntityQueryBuilder.executeUpdate(EntityQueryBuilder.java:115)
        at org.moqui.impl.entity.EntityValueImpl.createExtended(EntityValueImpl.java:102)
        at org.moqui.impl.entity.EntityValueBase.basicCreate(EntityValueBase.java:1245)
        at org.moqui.impl.entity.EntityValueBase.create(EntityValueBase.java:1211)
        ... 7 more

```

[moqui.log.txt](https://github.com/moqui/moqui-framework/files/922327/moqui.log.txt)

Any help would be appreciated. Thanks!

**Comment by jonesde** [comment]
Looking at the log there are more issues than just this one. The first error is "FOR UPDATE clause allowed only for DECLARE CURSOR" and looks like its getting a 50002 code from the DB. There is also a locking issue "Transaction (Process ID 53) was deadlocked on lock resources with another process and has been chosen as the deadlock victim. Rerun the transaction." with code 40001.

For the error you highlighted it looks like the JDBC driver doesn't like getting a String when it expects a number. The persistence level stuff has some disabled code to convert all fields to the proper type by configuration, but it is disabled for performance reasons as it adds a lot of overhead and rarely makes any difference... except in a case like this. In this case it's in the framework code and can be corrected, I'll look into that.

The only way to resolve the rest of these is to run against the database and research then work through each issue. Every database has its quirks. Most issues can be handled with configuration changes, but sometimes code changes are needed. In this case looking into transaction isolation and locking seems to be the big thing. For DB2 some adjustments were required, and even a configuration option to generate slightly different SQL.

That said, the work I did on DB2 was paid for by a client and for the next couple months at least I have way too much that needs to be done look into this more. If you find specific things that need to be changed I can look into it, and you might try posting in the Moqui Ecosystem group on LinkedIn (it has the biggest audience) to see if others have used MS SQL Server.

**Comment by jonesde** [comment]
I looked into the ArtifactHitBin and don't see anything there where a String is being used when a number should be... have you changed any of the type mapping configuration, perhaps changing 'id' from a NVARCHAR to a number? It would be nice to know with field or column was on, but the JDBC driver message isn't helpful there...

**Comment by joelschuster** [comment]
Thanks for the feedback!

I'm working off of the codebase without any changes except for the config file noted above.

However, looking into that particular error I did find this. It's an older post, but it seemed to possibly fit, depending on how MS wrote their driver: https://dertompson.com/2008/01/03/bigdecimal-and-jdbc-since-java-5-0/

**Comment by jonesde** [comment]
That's a mess... if it is really the case and MS hasn't fixed it since Java 5 then it's well overdue (Java 5 hit end of life in 2009). It might be possible to do something with this particular code, ie saving ArtifactHitBin, but BigDecimal is used for numbers all over the place as it is far more reliable than double/etc. If you find out what makes the SQLServer JDBC driver happy with BigDecimal or if a variation on the driver helps, please do comment here.

**Comment by joelschuster** [comment]
At this point I'm suspecting that the ArtifactHitBin is a symptom, not a cause.

Taking a look at just one single run, the start of the issue is here with a record lock conflict you mention above. I upped the transaction isolation, to no avail.

At this point I'm looking at EntityDataLoaderImpl.groovy.

```
 Creating table for moqui.basic.Enumeration pks: [enumId]
--- 2017-04-14 12:58:49.260 [   RequireNewTx] INFO                  org.moqui.impl.entity.EntityDbMeta []
 Created table ENUMERATION for entity moqui.basic.Enumeration in group transactional
--- 2017-04-14 13:00:29.281 [   RequireNewTx] WARN        org.moqui.impl.context.TransactionFacadeImpl []
 New transaction thread not terminated, in state RUNNABLE
--- 2017-04-14 13:00:31.244 [   RequireNewTx] WARN               org.moqui.impl.entity.EntityValueImpl []
 Error creating [moqui.basic.Enumeration: {enumTypeId=DataSourceType, lastUpdatedStamp=2017-04-14 12:58:49.205, enumId=DST_PURCHASED_DATA, description=Purchased Data}] tx a Bitronix Transaction with GTRID [3139322E3136382E312E3130320000015B6DD506F000000001], status=ACTIVE, 1 resource(s) enlisted (started Fri Apr 14 12:58:49 MDT 2017) con Group: transactional, Con: a ConnectionJavaProxy of a JdbcPooledConnection from datasource transactional_DS in state ACCESSIBLE with usage count 1 wrapping a JDBC LrcXAConnection on a JDBC LrcConnectionJavaProxy on ConnectionID:1 ClientConnectionId: 12a48cc7-f480-4886-82ec-9171ff155ab4 on a JDBC LrcConnectionJavaProxy on ConnectionID:1 ClientConnectionId: 12a48cc7-f480-4886-82ec-9171ff155ab4: com.microsoft.sqlserver.jdbc.SQLServerException: Transaction (Process ID 57) was deadlocked on lock resources with another process and has been chosen as the deadlock victim. Rerun the transaction.
--- 2017-04-14 13:00:31.270 [   RequireNewTx] WARN        org.moqui.impl.context.TransactionFacadeImpl []
 Transaction rollback. The rollback was originally caused by: Error loading entity data
org.xml.sax.SAXException: Error storing entity [moqui.basic.Enumeration] value (line 33): Error creating Enumeration {enumId=DST_PURCHASED_DATA}: record lock conflict found [40001]
	at org.moqui.impl.entity.EntityDataLoaderImpl$EntityXmlHandler.endElement(EntityDataLoaderImpl.groovy:695) ~[moqui_temp3966504317378599999WEB-INF_lib_moqui-framework-2.0.1-SNAPSHOT.jar.:2.0.1-SNAPSHOT]
	at org.apache.xerces.parsers.AbstractSAXParser.endElement(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
	at org.apache.xerces.parsers.AbstractXMLDocumentParser.emptyElement(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanStartElement(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
	at org.apache.xerces.parsers.XMLParser.parse(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
	at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
```


[moqui.log.txt](https://github.com/moqui/moqui-framework/files/922766/moqui.log.txt)
[MoquiActualConf.xml.txt](https://github.com/moqui/moqui-framework/files/922767/MoquiActualConf.xml.txt)

**Comment by jonesde** [comment]
What do you mean by 'upped the transaction isolation level'? For lock conflicts (especially deadlocks) you'd want to reduce the tx isolation because higher levels of isolation like RepeatableRead and Serializable may result in more full table or other bulk locks that will mess things up. There you get into the particulars of how the database does tx isolation, which was a trick with DB2 for example and required a workaround with custom SQL enabled by configuration (ie the for-update="FOR UPDATE WITH RS") setting. Something similar might be helpful for SQLServer.

**Comment by joelschuster** [comment]
I'm trying to understand why there's need for more than one transaction for just one user during initial load in the first place. I think I'm missing something.

**Comment by joelschuster** [comment]
Ok, I switched to a different driver; the jTds SQL Server driver. I got a different error out this time:

``` 
Created table ENUMERATION for entity moqui.basic.Enumeration in group transactional
--- 2017-04-14 13:20:20.966 [   RequireNewTx] WARN        org.moqui.impl.context.TransactionFacadeImpl []
 New transaction thread not terminated, in state RUNNABLE
--- 2017-04-14 13:20:23.974 [   RequireNewTx] WARN               org.moqui.impl.entity.EntityValueImpl []
 Error creating [moqui.basic.Enumeration: {enumTypeId=DataSourceType, lastUpdatedStamp=2017-04-14 13:18:40.898, enumId=DST_PURCHASED_DATA, description=Purchased Data}] tx a Bitronix Transaction with GTRID [3139322E3136382E312E3130320000015B6DE735FD00000001], status=ACTIVE, 1 resource(s) enlisted (started Fri Apr 14 13:18:40 MDT 2017) con Group: transactional, Con: a ConnectionJavaProxy of a JdbcPooledConnection from datasource transactional_DS in state ACCESSIBLE with usage count 1 wrapping a JDBC LrcXAConnection on a JDBC LrcConnectionJavaProxy on net.sourceforge.jtds.jdbc.JtdsConnection@7744a71f on a JDBC LrcConnectionJavaProxy on net.sourceforge.jtds.jdbc.JtdsConnection@7744a71f: java.sql.SQLException: The INSERT statement conflicted with the FOREIGN KEY constraint "EnumerationEnumerationType". The conflict occurred in database "moqui", table "dbo.ENUMERATION_TYPE", column 'ENUM_TYPE_ID'.
--- 2017-04-14 13:20:23.997 [   RequireNewTx] WARN        org.moqui.impl.context.TransactionFacadeImpl []
 Transaction rollback. The rollback was originally caused by: Error loading entity data
org.xml.sax.SAXException: Error storing entity [moqui.basic.Enumeration] value (line 33): Error creating Enumeration {enumId=DST_PURCHASED_DATA}: record already exists or related record does not exist [23000]
	at org.moqui.impl.entity.EntityDataLoaderImpl$EntityXmlHandler.endElement(EntityDataLoaderImpl.groovy:695) ~[moqui_temp3459261730401127600WEB-INF_lib_moqui-framework-2.0.1-SNAPSHOT.jar.:2.0.1-SNAPSHOT]
	at org.apache.xerces.parsers.AbstractSAXParser.endElement(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
	at org.apache.xerces.parsers.AbstractXMLDocumentParser.emptyElement(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanStartElement(Unknown Source) ~[xercesImpl-2.11.0.jar:?]
```

Does this help pinpoint the problem?

[moqui.log.txt](https://github.com/moqui/moqui-framework/files/922808/moqui.log.txt)

**Comment by jonesde** [comment]
No kidding, this is where the nuances of different databases get weird... I have no idea why SQLServer would report a lock conflict. You'd have to look at the database logs to see what happened and why records were locked, and what was locked. I've never seen anything like this on other databases, but until the cause is found who knows what's up. Usually you start running into issues like this when doing load testing or in tests like the concurrent order stuff in mantle-usl.

Even with a 'simple' data load there is stuff happening in separate transactions like the sequenced IDs and the async persistence of artifact hit data.

BTW, in general why are you even trying to use MS SQLServer? There are easy databases, and there are databases that are more pain than they are worth. There are also cheaper and more expensive databases. SQLServer is neither cheap nor easy so unless there is a business reason to use it then there's no point, and typically if there is a business reason to use then the business in question will have database support resources (internally or on a support agreement) that actually know something about the database... they really need to be involved for efforts like this.

In a way these days databases are simple, even for complex stuff, but that isn't true of the big commercial databases, not even their limited capacity 'free' versions.

**Comment by joelschuster** [comment]
See my above post with a different driver, it may inform the where's and why's. 

There is a vested interest in using MS SQL Server and there is a business reason for it. Unfortunately I can't get into the details.

**Comment by jonesde** [comment]
That latest error is really weird, the record it depends on is inserted just before it.

This could all be caused, for the earlier errors and this one, by creating tables on the fly as needed which Moqui does by default. You may not be able to do this with SQLServer (you can't with various databases, including DB2 and Postgres for example). For this set default-startup-add-missing="true" default-runtime-add-missing="false" on the database element (in MoquiDefaultConf.xml) or startup-add-missing="true" runtime-add-missing="false" on the datasource element.

Not knowing anything about SQLServer, but considering how many databases have issues with in-flight transactions and table creation or other meta data changes, this is likely the problem and the solution to it.

BTW, on my other comments the business reason doesn't matter so much, it's the availability of people and skill sets that matters and for things like this it is super helpful to work with someone who knows a lot about the database, at least to look into errors on the database side to get more information about what is happening.

**Comment by jonesde** [comment]
Let me know if the startup-add-missing works, I'll change the default settings for 'mssql' so that it does that without needing additional configuration (ie on the database element in MoquiDefaultConf.xml).

**Comment by joelschuster** [comment]
I got a bit further by running the load over and over. Finally, it now will fail here:

```
13:59:39.446  INFO RequireNewTx       o.moqui.i.e.EntityDataLoaderImpl Loaded 0 records from classpath://entity/OlapEntities.xml in 0.001s
13:59:39.447  INFO RequireNewTx       o.moqui.i.e.EntityDataLoaderImpl Loading entity data from classpath://entity/ResourceEntities.xml
13:59:39.447  INFO RequireNewTx       o.moqui.i.e.EntityDataLoaderImpl Loaded 0 records from classpath://entity/ResourceEntities.xml in 0s
13:59:39.448  INFO RequireNewTx       o.moqui.i.e.EntityDataLoaderImpl Loading entity data from classpath://entity/ScreenEntities.xml
13:59:39.507  INFO RequireNewTx       o.moqui.i.e.EntityDataLoaderImpl Loaded 40 records from classpath://entity/ScreenEntities.xml in 0.059s
13:59:39.509  INFO RequireNewTx       o.moqui.i.e.EntityDataLoaderImpl Loading entity data from classpath://entity/SecurityEntities.xml
13:59:39.570  WARN RequireNewTx      o.moqui.i.c.TransactionFacadeImpl Transaction rollback. The rollback was originally caused by: Error getting primary sequenced ID
org.moqui.impl.entity.EntitySqlException: Error finding one SequenceValueItem by seqName = moqui.entity.EntityAuditLog (java.lang.String) [S0002]
      at org.moqui.impl.entity.EntityFindBase.oneInternal(EntityFindBase.groovy:865) ~[moqui_temp6077450927071263258WEB-INF_lib_moqui-framework-2.0.1-SNAPSHOT.jar.:2.0.1-SNAPSHOT]
  at org.moqui.impl.entity.EntityFindBase.one(EntityFindBase.groovy:667) ~[moqui_temp6077450927071263258WEB-INF_lib_moqui-framework-2.0.1-SNAPSHOT.jar.:2.0.1-SNAPSHOT]
  at org.moqui.impl.entity.EntityFacadeImpl$_dbSequencedIdPrimary_closure11.doCall(EntityFacadeImpl.groovy:1789) ~[moqui_temp6077450927071263258WEB-INF_lib_moqui-framework-2.0.1-SNAPSHOT.jar.:2.0.1-SNAPSHOT]
  at org.moqui.impl.context.TransactionFacadeImpl.runUseOrBegin(TransactionFacadeImpl.groovy:166) ~[moqui_temp6077450927071263258WEB-INF_lib_moqui-framework-2.0.1-SNAPSHOT.jar.:2.0.1-SNAPSHOT]
 at org.moqui.impl.context.TransactionFacadeImpl$_runRequireNew_closure1.doCall(TransactionFacadeImpl.groovy:194) ~[moqui_temp6077450927071263258WEB-INF_lib_moqui-framework-2.0.1-SNAPSHOT.jar.:2.0.1-SNAPSHOT]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: Line 1: FOR UPDATE clause allowed only for DECLARE CURSOR.
  at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:217) ~[sqljdbc42.jar:?]
       at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1655) ~[sqljdbc42.jar:?]
      at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatement(SQLServerPreparedStatement.java:440) ~[sqljdbc42.jar:?]
  at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtExecCmd.doExecute(SQLServerPreparedStatement.java:385) ~[sqljdbc42.jar:?]
   at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7505) ~[sqljdbc42.jar:?]
```

[moqui.log.txt](https://github.com/moqui/moqui-framework/files/922870/moqui.log.txt)

**Comment by chunlinyao** [comment]
This may help
http://stackoverflow.com/questions/13717627/select-content-from-repository-where-id-for-update

And other reference.

Joel Schuster <notifications@github.com>于2017年4月15日周六 上午4:02写道：

**Comment by jonesde** [comment]
Hi @joelschuster , I noticed you closed this... did you get it working on SQL Server Express? What was the solution? I changed the startup and runtime add missing settings for the default 'mssql' database configuration, would be nice to know if that was sufficient.

Based on the comment from @chunlinyao I also added SelectMethod=Cursor to the default and example configurations in MoquiDefaultConf.xml (see commit #dd9bbf49).
